File: code/dart/test/plugin_host/official_plugins_basepath_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:modular_api/modular_api.dart';
import 'package:test/test.dart';

void main() {
  test('official operational endpoints resolve under the shared basePath', () async {
    final api = ModularApi(
      basePath: '/api',
      title: 'Plugin Ops',
      version: '1.0.0',
      metricsEnabled: true,
    );

    final server = await api.serve(port: 0);
    addTearDown(() => server.close(force: true));

    final client = HttpClient();
    addTearDown(client.close);

    final health = await _getJson(client, server.port, '/api/health');
    expect(health.$1, 200);
    expect((health.$2 as Map<String, dynamic>)['status'], 'pass');

    final metrics = await _getText(client, server.port, '/api/metrics');
    expect(metrics.$1, 200);
    expect(metrics.$2, contains('http_requests_total'));

    final openApiJson = await _getJson(client, server.port, '/api/openapi.json');
    expect(openApiJson.$1, 200);
    expect((openApiJson.$2 as Map<String, dynamic>)['openapi'], '3.0.0');

    final openApiYaml = await _getText(client, server.port, '/api/openapi.yaml');
    expect(openApiYaml.$1, 200);
    expect(openApiYaml.$2, contains('openapi: 3.0.0'));

    final docs = await _getText(client, server.port, '/api/docs');
    expect(docs.$1, 200);
    expect(docs.$2, contains('/api/openapi.json'));

    expect((await _getText(client, server.port, '/health')).$1, 404);
    expect((await _getText(client, server.port, '/metrics')).$1, 404);
    expect((await _getText(client, server.port, '/openapi.json')).$1, 404);
    expect((await _getText(client, server.port, '/openapi.yaml')).$1, 404);
    expect((await _getText(client, server.port, '/docs')).$1, 404);
  });
}

Future<(int, Object?)> _getJson(HttpClient client, int port, String path) async {
  final request = await client.getUrl(Uri.parse('http://127.0.0.1:$port$path'));
  final response = await request.close();
  final body = await utf8.decodeStream(response);
  return (response.statusCode, jsonDecode(body));
}

Future<(int, String)> _getText(HttpClient client, int port, String path) async {
  final request = await client.getUrl(Uri.parse('http://127.0.0.1:$port$path'));
  final response = await request.close();
  final body = await utf8.decodeStream(response);
  return (response.statusCode, body);
}

File: code/ts/test/plugin_host/official_plugins_basepath.test.ts
import { describe, expect, it } from 'vitest';
import request from 'supertest';

import { ModularApi } from '../../src';

describe('Official plugins under shared basePath', () => {
  it('serves operational endpoints under /api only', async () => {
    const api = new ModularApi({
      basePath: '/api',
      title: 'Plugin Ops',
      version: '1.0.0',
      metricsEnabled: true,
    });

    const server = await api.serve({ port: 0 });
    try {
      const health = await request(server).get('/api/health');
      expect(health.status).toBe(200);
      expect(health.body.status).toBe('pass');

      const metrics = await request(server).get('/api/metrics');
      expect(metrics.status).toBe(200);
      expect(metrics.text).toContain('http_requests_total');

      const openApiJson = await request(server).get('/api/openapi.json');
      expect(openApiJson.status).toBe(200);
      expect(openApiJson.body.openapi).toBe('3.0.0');

      const openApiYaml = await request(server).get('/api/openapi.yaml');
      expect(openApiYaml.status).toBe(200);
      expect(openApiYaml.text).toContain('openapi: 3.0.0');

      const docs = await request(server).get('/api/docs');
      expect(docs.status).toBe(200);
      expect(docs.text).toContain('/api/openapi.json');

      await request(server).get('/health').expect(404);
      await request(server).get('/metrics').expect(404);
      await request(server).get('/openapi.json').expect(404);
      await request(server).get('/openapi.yaml').expect(404);
      await request(server).get('/docs').expect(404);
    } finally {
      await new Promise<void>((resolve, reject) => {
        server.close((error) => {
          if (error) {
            reject(error);
            return;
          }

          resolve();
        });
      });
    }
  });
});